File: src/rmmz/plugin/core/parse/parse.ts
import type { Block, PluginStructBlock, PluginBodyBlock } from "./block";
import {
  filterSturctByLocale,
  findPluginBodyByLocale,
  splitBlock,
} from "./block";
import { flashCurrentItem, withTexts } from "./flashState";
import type { ParseState } from "./internalTypes";
import {
  handleBase,
  handleOption,
  handleOrderAfter,
  handleOrderBefore,
  handleValue,
} from "./state";
import { typeIsStruct } from "./struct";
import type {
  ParsedPlugin,
  PluginStructTokens,
  PluginCommandTokens,
  PluginMeta,
  KeywordEnum,
} from "./types";
import {
  KEYWORD_HELP,
  KEYWORD_PARAM,
  KEYWORD_TEXT,
  KEYWORD_DESC,
  KEYWORD_ARG,
  KEYWORD_STRUCT,
  KEYWORD_KIND,
  KEYWORD_DEFAULT,
  KEYWORD_ON,
  KEYWORD_OFF,
  KEYWORD_MIN,
  KEYWORD_MAX,
  KEYWORD_AUTHOR,
  KEYWORD_PLUGINDESC,
  KEYWORD_URL,
  KEYWORD_DECIMALS,
  KEYWORD_DIR,
  KEYWORD_PARENT,
} from "./types/keyword";

const bbb = (
  body: PluginBodyBlock,
  structs: readonly PluginStructBlock[]
): ParsedPlugin => {
  const finalState = parseBodyBlock(body);
  return {
    locale: body.locale,
    commands: finalState.commands,
    params: finalState.params,
    helpLines: finalState.helpLines,
    meta: finalState.meta,
    dependencies: finalState.dependencies,
    structs: structs
      .filter((s) => s.locale === body.locale)
      .map((s) => parseStructBlock(s)),
  };
};

export const parsePlugin2 = (text: string): ParsedPlugin[] => {
  const blocks: Block = splitBlock(text);
  return blocks.bodies.map((body): ParsedPlugin => bbb(body, blocks.structs));
};

export const parsePlugin = (
  text: string,
  locale: string = "en"
): ParsedPlugin => {
  const blocks: Block = splitBlock(text);
  const structs = filterSturctByLocale(blocks.structs, locale).map((s) =>
    parseStructBlock(s)
  );
  const body = findPluginBodyByLocale(blocks.bodies, locale);
  if (!body) {
    return {
      params: [],
      commands: [],
      meta: {},
      helpLines: [],
      dependencies: { base: [], orderBefore: [], orderAfter: [] },
      structs: structs,
    };
  }

  const finalState = parseBodyBlock(body);
  return {
    params: finalState.params,
    commands: finalState.commands,
    meta: finalState.meta,
    helpLines: finalState.helpLines,
    structs: structs,
    dependencies: finalState.dependencies,
  };
};

const parseStructBlock = (body: PluginStructBlock): PluginStructTokens => {
  const state = parseBodyBlock(body);
  return {
    name: body.struct,
    params: state.params,
  };
};

const parseBodyBlock = (body: PluginBodyBlock): ParseState => {
  const state = body.lines.reduce<ParseState>((acc, line) => {
    return parseBodyBlockLine(acc, line);
  }, makeParseState());
  return flashCurrentItem(state);
};

const makeParseState = (): ParseState => ({
  helpLines: [],
  params: [],
  commands: [],
  currentParam: null,
  currentCommand: null,
  currentContext: null,
  currentOption: null,
  dependencies: {
    base: [],
    orderBefore: [],
    orderAfter: [],
  },
  meta: {},
});

const parseBodyBlockLine = (
  acc: ParseState,
  line: string,
  table: Record<
    string,
    (state: ParseState, value: string) => ParseState
  > = KEYWORD_FUNC_TABLE
): ParseState => {
  // 先頭の'*'や空白を除去
  const trimmed = line.trimEnd().replace(/^[\*\s]*/, "");
  if (!trimmed.startsWith("@")) {
    // @タグがない場合はヘルプ行として追加するか無視
    if (acc.currentContext === KEYWORD_HELP) {
      return { ...acc, helpLines: acc.helpLines.concat(trimmed) };
    }
    return acc;
  }
  // @タグと値を抽出
  const match = trimmed.match(/^@(\S+)\s*(.*)$/);
  if (!match) {
    return acc;
  }
  const [, tag, value] = match;
  const fn = table[tag];
  if (fn) {
    return fn(acc, value.trim());
  }
  return acc;
};

const handleHelpContext = (oldstate: ParseState): ParseState => {
  const state = flashCurrentItem(oldstate);
  return {
    ...state,
    currentContext: KEYWORD_HELP,
  };
};

const handleParamContext = (
  oldstate: ParseState,
  value: string
): ParseState => {
  const state = flashCurrentItem(oldstate);
  // すでに同名のparamが存在する場合は新しいparamを作らない
  if (state.params.some((p) => p.name === value)) {
    return state;
  }
  return {
    ...state,
    currentContext: KEYWORD_PARAM,
    currentParam: { name: value, attr: {} },
  };
};

const handleText = (state: ParseState, value: string): ParseState => {
  if (state.currentParam) {
    return addParamField(state, KEYWORD_TEXT, value);
  }
  if (state.currentCommand) {
    if (!(KEYWORD_TEXT in state.currentCommand)) {
      return {
        ...state,
        currentCommand: {
          ...withTexts(state.currentCommand),
          command: state.currentCommand.command,
          args: state.currentCommand.args,
          [KEYWORD_TEXT]: value,
        },
      };
    }
  }
  return state;
};

const handleDesc = (state: ParseState, value: string): ParseState => {
  if (state.currentParam) {
    return addParamField(state, KEYWORD_DESC, value);
  }
  if (state.currentCommand) {
    return {
      ...state,
      currentCommand: {
        ...state.currentCommand,
        desc: value,
      },
    };
  }
  return state;
};

const handleCommandContext = (
  oldstate: ParseState,
  value: string
): ParseState => {
  const state: ParseState = flashCurrentItem(oldstate);
  if (state.commands.some((c) => c.command === value)) {
    // すでに同名のcommandが存在する場合は新しいcommandを作らない
    return state;
  }
  return {
    ...state,
    currentCommand: { command: value, args: [] },
    currentParam: null,
  };
};

const handleArgContext = (state: ParseState, value: string): ParseState => {
  if (!state.currentCommand) {
    return state;
  }
  if (!state.currentParam) {
    return {
      ...state,
      currentParam: {
        name: value,
        attr: {},
      },
    };
  }
  const argAddedCommand: PluginCommandTokens = {
    ...withTexts(state.currentCommand),
    command: state.currentCommand.command,
    args: state.currentCommand.args.concat(state.currentParam),
  };
  return {
    ...state,
    commands: state.commands,
    currentCommand: argAddedCommand,
    currentContext: KEYWORD_ARG,
    currentParam: {
      name: value,
      attr: {},
    },
  };
};

const handlerType = (state: ParseState, value: string): ParseState => {
  if (typeIsStruct(value)) {
    const structName = value.slice(7, -1);
    const addSturct = addParamField(state, KEYWORD_STRUCT, structName);
    return addParamField(addSturct, KEYWORD_KIND, KEYWORD_STRUCT);
  }
  if (state.currentParam) {
    return addParamField(state, KEYWORD_KIND, value);
  }
  return state;
};

const addParamField = (
  state: ParseState,
  key: string,
  value: string
): ParseState => {
  if (state.currentParam) {
    if (!(key in state.currentParam.attr)) {
      return {
        ...state,
        currentParam: {
          ...state.currentParam,
          attr: { ...state.currentParam.attr, [key]: value },
        },
      };
    }
  }
  return state;
};

const addMetaField = (
  state: ParseState,
  key: keyof PluginMeta,
  value: string
): ParseState => {
  return {
    ...state,
    meta: { [key]: value, ...state.meta },
  };
};

const KEYWORD_FUNC_TABLE = {
  param: handleParamContext,
  text: handleText,
  desc: handleDesc,
  command: handleCommandContext,
  arg: handleArgContext,
  help: handleHelpContext,
  option: handleOption,
  value: handleValue,
  type: handlerType,
  parent: (state, val) => addParamField(state, KEYWORD_PARENT, val),
  default: (state, value) => addParamField(state, KEYWORD_DEFAULT, value),
  on: (state, value) => addParamField(state, KEYWORD_ON, value),
  off: (state, value) => addParamField(state, KEYWORD_OFF, value),
  min: (state, value) => addParamField(state, KEYWORD_MIN, value),
  max: (state, value) => addParamField(state, KEYWORD_MAX, value),
  decimals: (state, value) => addParamField(state, KEYWORD_DECIMALS, value),
  dir: (state, value) => addParamField(state, KEYWORD_DIR, value),
  base: handleBase,
  orderAfter: handleOrderAfter,
  orderBefore: handleOrderBefore,
  author: (state, val) => addMetaField(state, KEYWORD_AUTHOR, val),
  plugindesc: (s, val) => addMetaField(s, KEYWORD_PLUGINDESC, val),
  url: (state, val) => addMetaField(state, KEYWORD_URL, val),
} as const satisfies {
  [K in KeywordEnum]?: (state: ParseState, value: string) => ParseState;
};

File: cs/parse.cs
using System.Text.RegularExpressions;
namespace RpgTypes{

    public sealed class ParsedPlugin
    {
        public string Locale { get; set; } = "";

        public List<PluginParamToken> Params { get; } = [];

        public List<PluginCommandToken> Commands { get; } = [];

        public List<PluginStructToken> Structs { get; } = [];

        public List<string> HelpLines { get; } = [];

        public PluginMeta Meta { get; set; } = new();

        public PluginDependencies Dependencies { get; set; } = new();
    }

    public sealed class PluginParamAttribute
    {
        public string? Kind { get; set; }

        public string? Struct { get; set; }

        public string? Text { get; set; }

        public string? Desc { get; set; }

        public string? Default { get; set; }

        public string? Min { get; set; }

        public string? Max { get; set; }

        public string? Parent { get; set; }

        public string? On { get; set; }

        public string? Off { get; set; }

        public string? Dir { get; set; }

        public string? Decimals { get; set; }
    }
    public readonly struct PluginDependencies
    {
        public List<string> Base { get; } = [];

        public List<string> OrderBefore { get; } = [];

        public List<string> OrderAfter { get; } = [];
    }
    public sealed class PluginStructToken
    {
        public required string Name { get; init; }

        public List<PluginParamToken> Params { get; } = [];
    }
    public sealed class PluginCommandToken
    {
        public required string Command { get; init; }

        public string? Text { get; set; }

        public string? Desc { get; set; }

        public List<PluginParamToken> Args { get; } = [];
    }

    public sealed class PluginParamToken
    {
        public required string Name { get; init; }

        public PluginParamAttribute Attr { get; } = new();

        public List<OptionItem> Options { get; } = [];
    }
    public sealed class OptionItem
    {
        public required string Option { get; init; }

        public required string Value { get; init; }
    }
    public sealed class PluginParser
    {
        private readonly ParseState _state = new();

        private readonly Dictionary<string, Action<string>> _handlers;

        public PluginParser()
        {
            _handlers = new Dictionary<string, Action<string>>(StringComparer.Ordinal)
            {
                ["help"] = HandleHelp,
                ["param"] = HandleParam,
                ["command"] = HandleCommand,
                ["arg"] = HandleArg,

                ["text"] = HandleText,
                ["desc"] = HandleDesc,
                ["type"] = HandleType,

                ["default"] = v => AddParamField(a => a.Default = v),
                ["min"] = v => AddParamField(a => a.Min = v),
                ["max"] = v => AddParamField(a => a.Max = v),
                ["on"] = v => AddParamField(a => a.On = v),
                ["off"] = v => AddParamField(a => a.Off = v),
                ["dir"] = v => AddParamField(a => a.Dir = v),
                ["parent"] = v => AddParamField(a => a.Parent = v),
                ["decimals"] = v => AddParamField(a => a.Decimals = v),
            };
        }

        public ParsedPlugin Parse(PluginBodyBlock body)
        {
            foreach (var line in body.Lines)
            {
                ParseLine(line);
            }

            FlushCurrentItem();

            return new ParsedPlugin
            {
                Locale = body.Locale,
                Params = _state.Params,
                Commands = _state.Commands,
                HelpLines = _state.HelpLines,
                Meta = _state.Meta,
                Dependencies = _state.Dependencies
            };
        }

        private void ParseLine(string line)
        {
            string trimmed = Regex.Replace(
                line.TrimEnd(),
                @"^[\*\s]*",
                ""
            );

            if (!trimmed.StartsWith("@"))
            {
                if (_state.CurrentContext == "help")
                {
                    _state.HelpLines.Add(trimmed);
                }
                return;
            }

            Match match = Regex.Match(trimmed, @"^@(\S+)\s*(.*)$");

            if (!match.Success)
                return;

            string tag = match.Groups[1].Value;
            string value = match.Groups[2].Value.Trim();

            if (_handlers.TryGetValue(tag, out var handler))
            {
                handler(value);
            }
        }
        private void HandleParam(string name)
        {
            FlushCurrentItem();

            if (_state.Params.Any(x => x.Name == name))
                return;

            _state.CurrentContext = "param";

            _state.CurrentParam = new PluginParamToken
            {
                Name = name
            };
        }

        private void HandleCommand(string name)
        {
            FlushCurrentItem();

            if (_state.Commands.Any(x => x.Command == name))
                return;

            _state.CurrentCommand = new PluginCommandToken
            {
                Command = name
            };
        }

        private void HandleArg(string name)
        {
            if (_state.CurrentCommand == null)
            {
                return;
            }

            if (_state.CurrentParam != null)
            {
                _state.CurrentCommand.Args.Add(_state.CurrentParam);
            }

            _state.CurrentParam = new PluginParamToken
            {
                Name = name
            };

            _state.CurrentContext = "arg";
        }

        private void HandleType(string value)
        {
            if (_state.CurrentParam == null){return;}

            if (value.StartsWith("struct<"))
            {
                bool array = value.EndsWith(">[]");

                string structName =
                    array
                        ? value.Substring(7, value.Length - 10)
                        : value.Substring(7, value.Length - 8);

                _state.CurrentParam.Attr.Struct = structName;
                _state.CurrentParam.Attr.Kind = array
                    ? "struct[]"
                    : "struct";

                return;
            }

            _state.CurrentParam.Attr.Kind = value;
        }

        private void HandleText(string value)
        {
            if (_state.CurrentParam != null)
            {
                _state.CurrentParam.Attr.Text ??= value;
                return;
            }

            if (_state.CurrentCommand != null)
            {
                _state.CurrentCommand.Text ??= value;
            }
        }
        private void HandleDesc(string value)
        {
            if (_state.CurrentParam != null)
            {
                _state.CurrentParam.Attr.Desc ??= value;
                return;
            }

            // if (_state.CurrentCommand != null)
            // {
            //     _state.CurrentCommand.Desc = value;
            // }
        }
        private void FlushCurrentItem()
        {
            FlushOptions();
            FlushCommand();
            FlushParam();
        }
        private void AddParamField(Action<PluginParamAttribute> setter)
        {
            if (_state.CurrentParam == null)
                {       return;}

            setter(_state.CurrentParam.Attr);
        }
        private void FlushCommand()
        {
            if (_state.CurrentCommand == null)
                return;

            if (_state.CurrentParam != null)
            {
                _state.CurrentCommand.Args.Add(_state.CurrentParam);
                _state.CurrentParam = null;
            }

            _state.Commands.Add(_state.CurrentCommand);

            _state.CurrentCommand = null;
            _state.CurrentContext = null;
}
        private void FlushParam()
        {
            if (_state.CurrentParam == null)
            {    return;}

            _state.Params.Add(_state.CurrentParam);

            _state.CurrentParam = null;
            _state.CurrentContext = null;
        }
    }
}